### DISCUSSION TITLE
[Question] Response timeouts for packages

### DISCUSSION THREAD

**Discussion by bozdoz** [npm/feedback#587] [CLI] [answered]
I saw as of npm-registry-fetch@4.0.3 that the default timeout for requests is 30s (https://github.com/npm/cli/commit/ea0ff56cb6ad4bbf208dd3cf39a35e07fc2dff16)

I think that *should* be fine. 

I'm curious about the endpoints that are being requested using `npm ci` for example.

My understanding is that the package-lock.json file should already have a "resolved" tarfile, and should just download that:

```json
    "typescript": {
      "version": "4.2.4",
      "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.2.4.tgz",
      "integrity": "sha512-V+evlYHZnQkaz8TRBuxTA92yZBPotr5H+WhQ7bD3hZUndx5tGOa1fuCgeSjxAzM1RiN5IzvadIXTVefuuwZCRg==",
      "dev": true
    }
```

But (given my network situation) this request can timeout; not because it's requesting the .tgz file, but because it's requesting https://registry.npmjs.org/typescript (which appears to be just a JSON list of versions?)

Why is this happening, if there is a resolved tarfile, and how could I prevent this/improve the performance of my installs?

**Comment by MylesBorins** [comment]
hey @bozdoz I've pinged @npm/cli to help answer this

**Reply by bozdoz** [reply]
Any idea how likely this is to be answered?

**Reply by bozdoz** [reply]
Oh well

**Comment by bozdoz** [comment, marked as answer]
Going to answer my own question.  I eventually figured this out by adding `DEBUG=*`, and a few other flags, and then looking into the files in the stack trace when the request times out.

### What is happening?

It's fetching a ["packument"](https://github.com/npm/pacote/blob/v9.5.12/README.md#--pacotepackumentspec-opts)

> Packument objects are general metadata about a project corresponding to registry metadata, and include version and dist-tag information about a package's available versions, rather than a specific version.

### Why is this happening?

Yes, there is a resolved tarfile; however, the "packument" will be fetched if the resolved value's registry url is not identical:

https://github.com/npm/pacote/blob/v9.5.12/lib/fetchers/registry/tarball.js#L18-L22

```js
if (
  opts.resolved &&
  // spec.type === 'version' &&
  opts.resolved.indexOf(registry) === 0
) {
// ...
}
```

In my case, *with a private registry*, there was a discrepancy between the registry urls.

My .npmrc:

```env
registry=https://artifactory.example.com/artifactory/npm/npm/
```

A sample "resolved" field:

```json
{
    "typescript": {
      "resolved": "https://artifactory.example.com:443/artifactory/npm/npm/typescript/-/typescript-4.2.4.tgz"
    }
}
```

This caused the "packument" to be fetched, which is huge, and caused the timeouts.

### How can I prevent this?

Change the registry in `.npmrc` to reflect the registry which is being resolved in the lockfile.

Thanks!
